packages/flutter/lib/src/foundation/persistent_hash_map.dart

```diff
@@ -0,0 +1,340 @@
+    return _bitCount(occupiedIndices & (bit - 1));
+  }
+
+  static _TrieNode _resolveCollision(int bitIndex, Object existingKey,
+      Object? existingValue, Object key, int hash, Object? value) {
+    final int existingKeyHash = existingKey.hashCode;
+    // Check if this is a full hash collision and use _HashCollisionNode
+    // in this case.
+    return (existingKeyHash == hash)
+        ? _HashCollisionNode.fromCollision(
+            hash, existingKey, existingValue, key, value)
+        : _CompressedNode.empty
+            .copyWith(bitIndex, existingKey, existingKeyHash, existingValue)
+            .copyWith(bitIndex, key, hash, value);
+  }
+}
+
+/// Trie node representing a full hash collision.
+///
+/// Stores a list of key/value pairs (where all keys have the same hash code).
+class _HashCollisionNode extends _TrieNode {
+  _HashCollisionNode(this.hash, this.keyValuePairs);
+
+  factory _HashCollisionNode.fromCollision(
+      int keyHash, Object keyA, Object? valueA, Object keyB, Object? valueB) {
+    final List<Object?> list = List<Object?>.filled(4, null);
+    list[0] = keyA;
+    list[1] = valueA;
+    list[2] = keyB;
+    list[3] = valueB;
+    return _HashCollisionNode(keyHash, list);
+  }
+
+  final int hash;
+  final List<Object?> keyValuePairs;
+
+  @override
+  _TrieNode copyWith(int bitIndex, Object key, int keyHash, Object? val) {
+    // Is this another full hash collision?
+    if (keyHash == hash) {
+      final int index = _indexOf(key);
+      if (index != -1) {
+        return identical(keyValuePairs[index + 1], val)
+            ? this
+            : _HashCollisionNode(
+                keyHash, _copy(keyValuePairs)..[index + 1] = val);
+      }
+      final int length = keyValuePairs.length;
+      final List<Object?> newArray = List<Object?>.filled(length + 2, null);
+      for (int i = 0; i < length; i++) {
+        newArray[i] = keyValuePairs[i];
+      }
+      newArray[length] = key;
+      newArray[length + 1] = val;
+      return _HashCollisionNode(keyHash, newArray);
+    }
+
+    // Not a full hash collision, need to introduce a _CompressedNode which
+    // uses previously unused bits.
+    return _CompressedNode.single(bitIndex, hash, this)
+        .copyWith(bitIndex, key, keyHash, val);
+  }
+
+  @override
+  Object? lookup(int bitIndex, Object key, int keyHash) {
+    final int index = _indexOf(key);
+    return index < 0 ? null : keyValuePairs[index + 1];
+  }
+
+  int _indexOf(Object key) {
+    final int length = keyValuePairs.length;
+    for (int i = 0; i < length; i += 2) {
+      if (key == keyValuePairs[i]) {
+        return i;
+      }
+    }
+    return -1;
+  }
+}
+
+@pragma('vm:prefer-inline')
+int _bitCount(int n) {
+  n = n - ((n >> 1) & 0x55555555);
+  n = (n & 0x33333333) + ((n >> 2) & 0x33333333);
+  n = (n + (n >> 4)) & 0x0F0F0F0F;
+  n = n + (n >> 8);
+  n = n + (n >> 16);
+  return n & 0x0000003F;
+}
+
+List<Object?> _copy(List<Object?> array) {
```

Comment:
nit: remove this function and just replace it with `List<Object?>.from(array)` everywhere.

Reply:
I need to benchmark. I would not be surprised if it is slower to use `from` due to polymorphism inside `from` - it takes `Iterable` and is actually called from various places.

Reply:
Yeah. `from` and `of` are 2x slower.

Reply:
We should really fix `from` and `of` then.

I would expect `from` to be a little slower because it involves more casting, but I thought `of` was supposed to basically be this (or maybe I have these reversed?)

Reply:
Filled one more: https://github.com/dart-lang/sdk/issues/49408